package com.james.store.controllers;

import java.util.Map;
import java.util.UUID;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import com.james.store.dtos.AddItemToCartRequest;
import com.james.store.dtos.CartDto;
import com.james.store.dtos.CartItemDto;
import com.james.store.dtos.UpdateCartItemDto;
import com.james.store.exceptions.CartNotFoundException;
import com.james.store.exceptions.ProductNotFoundException;
import com.james.store.services.CartService;

import jakarta.validation.Valid;
import lombok.AllArgsConstructor;

@RestController
@AllArgsConstructor
@RequestMapping("/carts")
public class CartController {
  private final CartService cartService;

  @PostMapping
  @ResponseStatus(HttpStatus.CREATED)
  public ResponseEntity<CartDto> createCart() {
    var cart = cartService.createCart();
    return ResponseEntity.ok(cart);
  }
  
  @PostMapping("/{cartId}/items")
  @ResponseStatus(HttpStatus.CREATED)
  public ResponseEntity<CartItemDto> addToCart(@PathVariable(name = "cartId") UUID cartId, @Valid @RequestBody AddItemToCartRequest request) {
    var cartItemDto = cartService.addToCart(cartId, request.getProductId());
    return ResponseEntity.ok(cartItemDto);
  }

  @GetMapping("/{cartId}")
  @ResponseStatus(HttpStatus.OK)
  public ResponseEntity<CartDto> getCart(@PathVariable(name = "cartId") UUID cartId) {
    var cartDto = cartService.getCart(cartId);
    return ResponseEntity.ok(cartDto);
  }
  
  @PutMapping("/{cartId}/items/{productId}")
  public ResponseEntity<?> updateCartItem(
    @PathVariable(name = "cartId") UUID cartId,
    @PathVariable(name = "productId") Long productId,
    @Valid @RequestBody UpdateCartItemDto request
  ) {
    var cartItemDto = cartService.updateItem(cartId, productId, request.getQuantity());
    return ResponseEntity.ok(cartItemDto);
  }

  @DeleteMapping("/{cartId}/items/{productId}")
  public ResponseEntity<?> removeProductFromCart(
    @PathVariable(name = "cartId") UUID cartId,
    @PathVariable(name = "productId") Long productId
  ) {
    cartService.removeItem(cartId, productId);
    return ResponseEntity.noContent().build();
  }

  @DeleteMapping("/{cartId}/items")
  public ResponseEntity<Void> clearCart(
    @PathVariable(name = "cartId") UUID cartId
  ) {
    cartService.clear(cartId);
    return ResponseEntity.noContent().build();
  }
  
  @ExceptionHandler(CartNotFoundException.class)
  public ResponseEntity<Map<String, String>> handleCartNotFound() {
    return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Cart not found"));
  }

  @ExceptionHandler(ProductNotFoundException.class)
  public ResponseEntity<Map<String, String>> handleProductNotFound() {
    return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("error", "Product not found"));
  }

}
